src/main/java/com/puppycrawl/tools/checkstyle/doclets/TokenTypesDoclet.java

```diff
@@ -19,148 +19,237 @@
+     * Checks that the DetailAST is a {@code public} {@code static} {@code final} {@code int} field.
+     * @param ast to process
+     * @return {@code true} if matches, {@code false} otherwise.
      */
-    public static int optionLength(String option) {
-        int length = 0;
-        if (DEST_FILE_OPT.equals(option)) {
-            length = 2;
+    private static boolean isPublicStaticFinalIntField(DetailAST ast) {
+        boolean result = ast.getType() == TokenTypes.VARIABLE_DEF;
+        if (result) {
+            final DetailAST type = ast.findFirstToken(TokenTypes.TYPE);
+            result = type.findFirstToken(TokenTypes.LITERAL_INT) != null;
+            if (result) {
+                final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
+                result = modifiers.findFirstToken(TokenTypes.LITERAL_PUBLIC) != null
+                    && modifiers.findFirstToken(TokenTypes.LITERAL_STATIC) != null
+                    && modifiers.findFirstToken(TokenTypes.FINAL) != null;
+            }
         }
-        return length;
+        return result;
     }
 
     /**
-     * Checks that only valid options was specified.
-     * @param options all parsed options
-     * @param reporter the reporter to report errors.
-     * @return true if only valid options was specified
+     * Extracts the name of an ast.
+     * @param ast to extract the name
+     * @return the text content of the inner {@code TokenTypes.IDENT} node
      */
-    public static boolean checkOptions(String[][] options, DocErrorReporter reporter) {
-        boolean foundDestFileOption = false;
-        boolean onlyOneDestFileOption = true;
-        for (final String[] opt : options) {
-            if (DEST_FILE_OPT.equals(opt[0])) {
-                if (foundDestFileOption) {
-                    reporter.printError("Only one -destfile option allowed.");
-                    onlyOneDestFileOption = false;
-                    break;
+    private static String getName(DetailAST ast) {
+        return ast.findFirstToken(TokenTypes.IDENT).getText();
+    }
+
+    /**
+     * Extracts the first sentence of an ast javadoc comment.
+     * @param ast to extract the first sentence
+     * @return the first sentence of the inner {@code TokenTypes.BLOCK_COMMENT_BEGIN} node
+     *     or {@code null} if the first sentence is absent or malformed (does not end with period)
+     */
+    private static String getFirstJavadocSentence(DetailAST ast) {
+        String firstSentence = null;
+        final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
+        for (DetailAST comment = modifiers.getFirstChild(); comment != null;
+                comment = comment.getNextSibling()) {
+            if (comment.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
+                    && JavadocUtils.isJavadocComment(comment)) {
+                final String plainJavadocText = JAVADOC_MULTILINE_TO_SINGLELINE_PATTERN
```

Comment:
If we are going through the hassle of using our Grammar, why do we need a RegExp parser for the javadocs and can't just use our Javadoc grammar?
Regular expressions are known to be bad for javadocs which is why we use the grammar.

Reply:
The javadoc parser is designed for other purpose. He does not have the tag "first sentence".
Here is what it looks like:
```
    |   |   |   |   `--JAVADOC -> JAVADOC [15:7]
    |   |   |   |       |--NEWLINE -> \n [15:7]
    |   |   |   |       |--LEADING_ASTERISK ->      * [16:0]
    |   |   |   |       |--TEXT ->  '&amp;' symbol when used in a generic upper or lower bounds constrain [16:6]
    |   |   |   |       |--NEWLINE -> \n [16:76]
    |   |   |   |       |--LEADING_ASTERISK ->      * [17:0]
    |   |   |   |       |--TEXT ->  e.g.&nbsp; [17:6]
    |   |   |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [17:17]
    |   |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [17:17]
    |   |   |   |       |   |--CODE_LITERAL -> @code [17:18]
    |   |   |   |       |   |--WS ->   [17:23]
    |   |   |   |       |   |--TEXT -> Comparable&lt;? extends Serializable, CharSequence&gt; [17:24]
    |   |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [17:79]
    |   |   |   |       |--TEXT -> . [17:80]
    |   |   |   |       |--NEWLINE -> \n [17:81]
    |   |   |   |       |--TEXT ->       [18:0]
    |   |   |   |       `--EOF -> <EOF> [18:5]
```

We need to do much more then a simple regexp match.
I'll give it a try, but i'm sceptical about it.

Reply:
@pbludov See `SummaryJavadocCheck`. It does the same thing and uses our Javadoc parser.

> The javadoc parser is designed for other purpose

The parser is designed for all our Javadoc needs. There is no other purpose for it then that.

Reply:
Really? The code you are complain about was taken exactly from this class.
It does not use Javdoc parser.

Reply:
Done. Now the code is more reliable. And slow.

Reply:
> The code you are complain about was taken exactly from this class.

It extends `AbstractJavadocCheck` which means it does use the Javadoc parser not regular expressions.